index.js

```diff
@@ -0,0 +1,109 @@
+
+// Store Slack access token in a .env file in the root folder
+// You can ping a JDSB admin on Slack for the OAuth token, or request to become a collaborator on the app.
+// The token for an app can be found by selecting the app from api.slack.com/apps and navigating to "OAuth & Permissions" page after it has been installed to a workspace.
+const accessToken = process.env.SLACK_ACCESS_TOKEN;
+
+async function getConversationId(conversation) {
+
+  // Trim the pound sign if it was provided in the conversation argument
+  if (conversation.startsWith('#')) {
+    conversation = conversation.slice(1)
+  }
+
+  // Destructuring to pull out the channels result from the response
+  const { data: { channels } } = await axios.get(`https://slack.com/api/conversations.list?token=${accessToken}&pretty=1`)
+
+  // Find the channel with matching conversation name and pull out its ID
+  const conversationId = channels.find(chan => chan.name === conversation).id
+
+  return conversationId
+}
+
+async function getConversationMembers(conversationId) {
+
+  async function queryForMembers(membersCollection = [], nextCursor = '') {
+    let allConversationMembers = [];
+
+    // Base case: members exist, no next cursor, we've retrieved all the members
+    if (membersCollection.length > 0 && !nextCursor) {
+      return membersCollection;
+    }
+
+    try {
+      const response = await axios.get(
+        `https://slack.com/api/conversations.members?token=${accessToken}&channel=${conversationId}${nextCursor}`
+      );
+
+      // Looks ugly, but pulling out members and next_cursor
+      let {
+        data: {
+          members,
+          response_metadata: { next_cursor }
+        }
+      } = response;
+
+      // Cursor seems to always end with =
+      // Replace with URL encoding, %3D
+      if (next_cursor) {
+        next_cursor = `&cursor=${next_cursor.replace('=', '%3D')}`
+      }
+
+      // Combine member grouping with previous
+      allConversationMembers = [...members, ...membersCollection]
+
+      return await queryForMembers(allConversationMembers, next_cursor)
+    }
+    catch (err) {
+      console.log(err);
+    }
+
+  }
+
+  // Kick off recursive call
+  return await queryForMembers()
+}
+
+async function inviteMembers(sourceChannel, targetChannel) {
+
+  // Basic check for 2 arguments
+  if (process.argv.length !== 4) {
+    return console.log("Please supply a source channel and a target channel")
+  }
+
+  // Retrieve the ID of the source and target channels for member access
+  const sourceId = await getConversationId(sourceChannel)
+  const targetId = await getConversationId(targetChannel)
+
+  // Retrieve member lists for comparison
+  const sourceMembers = await getConversationMembers(sourceId)
+  const targetMembers = await getConversationMembers(targetId)
```

Comment:
I think, this await and the ones above it are not taking advantage concurrency. Basically, each one is waiting on the last promise to finish before running its own code. You might as well get rid of the async function and call it synchronously.

IRC, you need to encapsulate these function call in yet another promise to run concurrently.

Reference: https://tutorialzine.com/2017/07/javascript-async-await-explained

Reply:
Ah, I see what you're saying. 

There is a better way regarding concurrency, yeah -- would have to refresh  my memory on best practice there -- but the actual purpose of awaiting at all here is for the code below, so `getConversationMembers()` doesn't run until the IDs have returned, and so we don't filter into `membersToInvite` until the member lists have returned.

Reply:
@bhurstGH Yeah, your membersToInvite is your synchronization point, but the above four should be able to run concurrently...they don't depend on one another.

Reply:
@bhurstGH and I agree to save this for the migration to a Slack App.